# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

- **Consider a significance level of 5% for all tests.**

In [2]:
# Relevant libraries 
import pandas as pd
import numpy as np
import scipy.stats as stats


# Challenge 1 - Independent Sample T-tests

In this challenge, we will be using the Pokemon dataset. Before applying statistical methods to this data, let's first examine the data.

To load the data, run the code below.

In [4]:
pokemon = pd.read_csv(r"C:\Users\HP\Desktop\Ironhack\Labs\lab-two-sample-hyp-test\pokemon.csv")

Let's start off by looking at the `head` function in the cell below.

In [6]:
pokemon.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

In [7]:
pokemon.shape

(800, 13)

In [8]:
pokemon.isna().sum()

#               0
Name            0
Type 1          0
Type 2        386
Total           0
HP              0
Attack          0
Defense         0
Sp. Atk         0
Sp. Def         0
Speed           0
Generation      0
Legendary       0
dtype: int64

The first thing we would like to do is compare the legendary Pokemon to the regular Pokemon. To do this, we should examine the data further. What is the count of legendary vs. non legendary Pokemons?

In [10]:
pokemon["Legendary"].value_counts()

Legendary
False    735
True      65
Name: count, dtype: int64

Compute the mean and standard deviation of the total points for both legendary and non-legendary Pokemon.

In [12]:
df_legendary = round(pokemon.groupby(["Legendary"])["Total"].agg(["mean", "std"]), 2)
df_legendary

mean     std
Legendary                
False      417.21  106.76
True       637.38   60.94

The computation of the mean might give us a clue regarding how the statistical test may turn out; However, it certainly does not prove whether there is a significant difference between the two groups.

In the cell below, use the `ttest_ind` function in `scipy.stats` to compare the the total points for legendary and non-legendary Pokemon. Since we do not have any information about the population, assume the variances are not equal.

#### The Hypotheses
- **Null Hypothesis (H₀)**: The mean total points for non-legendary pokemon is equal to the mean total points for legendary pokemon.

- **Alternative Hypothesis (H₁)**: The mean total points for non-legendary pokemon is not equal to the mean total points for legendary pokemon.

In [15]:
#Definig the samples
legendary_points = pokemon[pokemon["Legendary"] == True]["Total"]
non_legendary_points = pokemon[pokemon["Legendary"] == False]["Total"]

In [16]:
legendary_points.describe()

count     65.000000
mean     637.384615
std       60.937389
min      580.000000
25%      580.000000
50%      600.000000
75%      680.000000
max      780.000000
Name: Total, dtype: float64

In [17]:
# Performing two-sample t-test for independent samples
t_stat, p_value = stats.ttest_ind(legendary_points, non_legendary_points, equal_var=False)
print(f"Test Statistic (t): {t_stat:.2f}")
print(f"P-Value: {p_value:.5f}")
print()

Test Statistic (t): 25.83
P-Value: 0.00000



What do you conclude from this test? Write your conclusions below.

- Given our significance level of 5% (alpha = 0.05) and a p-value equal to 0, we may reject the Null Hypothesis. In other words, there is sufficient evidence to conclude that there is a significant difference between the total points' means of Legendary and Non-legendary pokemon.

How about we try to compare the different types of pokemon? In the cell below, list the types of Pokemon from column `Type 1` and the count of each type.

In [21]:
pokemon["Type 1"].value_counts()

Type 1
Water       112
Normal       98
Grass        70
Bug          69
Psychic      57
Fire         52
Electric     44
Rock         44
Dragon       32
Ground       32
Ghost        32
Dark         31
Poison       28
Steel        27
Fighting     27
Ice          24
Fairy        17
Flying        4
Name: count, dtype: int64

Since water is the largest group of Pokemon, compare the mean and standard deviation of water Pokemon to all other Pokemon.

In [66]:
#Filtering by water type and non-water type
water_pokemon = pokemon[pokemon["Type 1"] == "Water"]
non_water_pokemon = pokemon[pokemon["Type 1"] != "Water"]

In [72]:
water_pokemon_mean = round(water_pokemon.Total.mean(), 2)
print(f"Water pokemon type have mean of {water_pokemon_mean} total points")

Water pokemon type have mean of 430.46 total points


In [76]:
non_water_pokemon_mean = round(non_water_pokemon.Total.mean(), 2)
print(f"Non-water pokemon type have mean of {non_water_pokemon_mean} total points")

Non-water pokemon type have mean of 435.86 total points


In [78]:
water_pokemon_std = round(water_pokemon.Total.std(), 2)
print(f"Water pokemon total points distribution has a standard deviation equal to {water_pokemon_std}")

Water pokemon total points distribution has a standard deviation equal to 113.19


In [80]:
non_water_pokemon_std = round(non_water_pokemon.Total.std(), 2)
print(f"Non-Water pokemon total points distribution has a standard deviation equal to {non_water_pokemon_std}")

Non-Water pokemon total points distribution has a standard deviation equal to 121.09


Perform a hypothesis test comparing the mean of total points for water Pokemon to all non-water Pokemon. Assume the variances are equal. 

#### The Hypotheses
- **Null Hypothesis (H₀)**: The mean total points for non-water type pokemon is equal to the mean total points for water type pokemon.

- **Alternative Hypothesis (H₁)**: The mean total points for non-water type pokemon is not equal to the mean total points for water type pokemon.

In [116]:
#Defining our samples:
water_pokemon_points = pokemon[pokemon["Type 1"] == "Water"]["Total"]
non_water_pokemon_points = pokemon[pokemon["Type 1"] != "Water"]["Total"]

In [118]:
# Performing two-sample t-test for independent samples
t_stat, p_value = stats.ttest_ind(water_pokemon_points, non_water_pokemon_points, equal_var=True)
print(f"Test Statistic (t): {t_stat:.2f}")
print(f"P-Value: {p_value:.5f}")
print()

Test Statistic (t): -0.44
P-Value: 0.65871



Write your conclusion below.

- Given our significance level of 5% (alpha = 0.05) and a p-value equal to 0,65871, we may conclude that there is no sufficient evidence to reject the Null Hypothesis, in other words, there is no sufficient evidence to state that there is a significant difference between the total points' means of Water type and Non-water type pokemon.


# Challenge 2 - Matched Pairs Test

In this challenge we will compare dependent samples of data describing our Pokemon. Our goal is to see whether there is a significant difference between each Pokemon's defense and attack scores. Our hypothesis is that the defense and attack scores are equal. In the cell below, import the `ttest_rel` function from `scipy.stats` and compare the two columns to see if there is a statistically significant difference between them.

#### The Hypotheses
- **Null Hypothesis (H₀)**: The mean of pokemon attack points is equal to the mean of pokemon defense points.

- **Alternative Hypothesis (H₁)**: The mean of pokemon attack points is not equal to the mean of pokemon defense points.

In [94]:
#Definig our samples:
attack_points = pokemon["Attack"]
defense_points = pokemon["Defense"]

In [96]:
# Performing two-sample t-test for dependent samples
t_stat, p_value = stats.ttest_rel(attack_points, defense_points)
print(f"Test Statistic (t): {t_stat:.2f}")
print(f"P-Value: {p_value:.5f}")
print()

Test Statistic (t): 4.33
P-Value: 0.00002



Describe the results of the test in the cell below.

- Given our significance level of 5% (alpha = 0.05) and a p-value close to 0, we may reject the Null Hypothesis. In other words, there is sufficient evidence to conclude that there is a significant difference between the means of pokemon attack and defense points.

We are also curious about whether therer is a significant difference between the mean of special defense and the mean of special attack. Perform the hypothesis test in the cell below. 

#### The Hypotheses
- **Null Hypothesis (H₀)**: The mean of pokemon special attack points is equal to the mean of pokemon special defense points.

- **Alternative Hypothesis (H₁)**: The mean of pokemon special attack points is not equal to the mean of pokemon special defense points.


In [101]:
#Definig our samples:
sp_attack_points = pokemon["Sp. Atk"]
sp_defense_points = pokemon["Sp. Def"]

In [103]:
# Performing two-sample t-test for dependent samples
t_stat, p_value = stats.ttest_rel(sp_attack_points, sp_defense_points)
print(f"Test Statistic (t): {t_stat:.2f}")
print(f"P-Value: {p_value:.5f}")
print()

Test Statistic (t): 0.85
P-Value: 0.39337



Describe the results of the test in the cell below.

- Given our significance level of 5% (alpha = 0.05) and a p-value equal to 0.39337, we may conclude that there is no sufficient evidence to reject the Null Hypothesis, in other words, there is no sufficient evidence to state that there is a significant difference between the means of pokemon special attack and defense points.



As you may recall, a two sample matched pairs test can also be expressed as a one sample test of the difference between the two dependent columns.

Import the `ttest_1samp` function and perform a one sample t-test of the difference between defense and attack. Test the hypothesis that the difference between the means is zero. Confirm that the results of the test are the same.

#### The Hypotheses
- **Null Hypothesis (H₀)**: The difference between the means of pokemon special attack and defense points is zero.

- **Alternative Hypothesis (H₁)**: the difference between the means of pokemon special attack and defense points is not zero.

In [109]:
#Definig our sample:
dif_sp_points = sp_attack_points - sp_defense_points  
    

In [113]:
# Performing two_tail t-test
t_stat, p_value = stats.ttest_1samp(dif_sp_points, 0) #our hypothysed population mean is 0 given that we are evaluating if the difference is 0
print(f"Test Statistic (t): {t_stat:.2f}")
print(f"P-Value: {p_value:.5f}")
print()

Test Statistic (t): 0.85
P-Value: 0.39337



- We can spot that the test results are same and so are the conclusions.

# Bonus Challenge - The Chi-Square Test

The Chi-Square test is used to determine whether there is a statistically significant difference in frequencies. In other words, we are testing whether there is a relationship between categorical variables or rather when the variables are independent. This test is an alternative to Fisher's exact test and is used in scenarios where sample sizes are larger. However, with a large enough sample size, both tests produce similar results. Read more about the Chi Squared test [here](https://en.wikipedia.org/wiki/Chi-squared_test).

In the cell below, create a contingency table using `pd.crosstab` comparing whether a Pokemon is legenadary or not and whether the Type 1 of a Pokemon is water or not.

In [122]:
df_pokemon = pokemon.copy()

In [126]:
df_pokemon["Water Type"] = df_pokemon["Type 1"]

In [134]:
for x in df_pokemon["Type 1"]:
    if x == "Water":
        df_pokemon["Water Type"].replace({x:True}, inplace=True)
    else:
        df_pokemon["Water Type"].replace({x:False}, inplace = True)
        

C:\Users\HP\AppData\Local\Temp\ipykernel_16420\611807651.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_pokemon["Water Type"].replace({x:False}, inplace = True)
C:\Users\HP\AppData\Local\Temp\ipykernel_16420\611807651.py:5: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_pokemon["Water Type

In [136]:
df_pokemon["Water Type"].value_counts()

Water Type
False    688
True     112
Name: count, dtype: int64

In [138]:
crosstab_result = pd.crosstab(df_pokemon["Legendary"], df_pokemon["Water Type"])
crosstab_result

Water Type  False  True 
Legendary               
False         627    108
True           61      4

Perform a chi-squared test using the `chi2_contingency` function in `scipy.stats`. You can read the documentation of the function [here](https://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.stats.chi2_contingency.html).

In [142]:
from scipy.stats import chi2_contingency

In [148]:
# Chi-square test for "Legendary" and "Water Type"
chi2_stats, chi2_pvalue, _, _ = chi2_contingency(crosstab_result)
print(f"chi2_pvalue: {chi2_pvalue}")

chi2_pvalue: 0.08625467249550949


Based on a 95% confidence, should we reject the null hypothesis?

- The null hypothesis of the Chi-square test is that the two categorical variables, Legendary and Water Type, are independent. Given the p-value of 0.08, higher than our significance level of 0.5, we may conclude that there is no sufficient evidence to reject the Null Hypothesis. In other words, there is no sufficient evidence to state that Legendary and Water Type are not independent variabes.